src/QuickCheckVEngine/RVFI_DII/RVFI.hs

```diff
@@ -73,175 +84,404 @@ type RV_RegIdx = Word8
 rvfiGetFromString _           = Nothing
 
-instance Binary RVFI_Packet where
-  put pkt =    putWord8    (rvfi_intr pkt)
-            <> putWord8    (rvfi_halt pkt)
-            <> putWord8    (rvfi_trap pkt)
-            <> putWord8    (rvfi_rd_addr pkt)
-            <> putWord8    (rvfi_rs2_addr pkt)
-            <> putWord8    (rvfi_rs1_addr pkt)
-            <> putWord8    (rvfi_mem_wmask pkt)
-            <> putWord8    (rvfi_mem_rmask pkt)
-            <> putWord64be (rvfi_mem_wdata pkt)
-            <> putWord64be (rvfi_mem_rdata pkt)
-            <> putWord64be (rvfi_mem_addr pkt)
-            <> putWord64be (rvfi_rd_wdata pkt)
-            <> putWord64be (rvfi_rs2_rdata pkt)
-            <> putWord64be (rvfi_rs1_rdata pkt)
-            <> putWord64be (rvfi_insn pkt)
-            <> putWord64be (rvfi_pc_wdata pkt)
-            <> putWord64be (rvfi_pc_rdata pkt)
-            <> putWord64be (rvfi_order pkt)
-  get = do intr      <- getWord8
-           halt      <- getWord8
-           trap      <- getWord8
-           rd_addr   <- getWord8
-           rs2_addr  <- getWord8
-           rs1_addr  <- getWord8
-           mem_wmask <- getWord8
-           mem_rmask <- getWord8
-           mem_wdata <- getWord64be
-           mem_rdata <- getWord64be
-           mem_addr  <- getWord64be
-           rd_wdata  <- getWord64be
-           rs2_rdata <- getWord64be
-           rs1_rdata <- getWord64be
-           insn      <- getWord64be
-           pc_wdata  <- getWord64be
-           pc_rdata  <- getWord64be
-           order     <- getWord64be
-           return $ RVFI_Packet {
-               rvfi_valid = 1
-             , rvfi_order = order
-             , rvfi_insn  = insn
-             , rvfi_trap  = trap
-             , rvfi_halt  = halt
-             , rvfi_intr  = intr
-             --, rvfi_mode = _
-             --, rvfi_ixl  = _
-             , rvfi_rs1_addr  = rs1_addr
-             , rvfi_rs2_addr  = rs2_addr
-             , rvfi_rs1_rdata = rs1_rdata
-             , rvfi_rs2_rdata = rs2_rdata
-             , rvfi_rd_addr   = rd_addr
-             , rvfi_rd_wdata  = rd_wdata
-             , rvfi_pc_rdata  = pc_rdata
-             , rvfi_pc_wdata  = pc_wdata
-             , rvfi_mem_addr  = mem_addr
-             , rvfi_mem_rmask = mem_rmask
-             , rvfi_mem_wmask = mem_wmask
-             , rvfi_mem_rdata = mem_rdata
-             , rvfi_mem_wdata = mem_wdata
-             }
+rvfi_rd_wdata_or_zero :: RVFI_Packet -> Word64
+rvfi_rd_wdata_or_zero x = rvfi_rd_wdata (fromMaybe rvfiEmptyIntData (rvfi_int_data x))
+
+data RVFI_IntData = RVFI_IntData
+  { rvfi_rs1_addr :: {-# UNPACK #-} !RV_RegIdx,
+    rvfi_rs2_addr :: {-# UNPACK #-} !RV_RegIdx,
+    rvfi_rs1_rdata :: {-# UNPACK #-} !RV_WordXLEN,
+    rvfi_rs2_rdata :: {-# UNPACK #-} !RV_WordXLEN,
+    rvfi_rd_addr :: {-# UNPACK #-} !RV_RegIdx,
+    rvfi_rd_wdata :: {-# UNPACK #-} !RV_WordXLEN
+  }
+  deriving (Show)
```

Comment:
Why use strict fields here (`RVFI_IntData` and `RVFI_MemAccessData`)? (not suggesting it is a bad idea, just genuinely wondering... as in, maybe you'd want them in the `RVFI_Packet` record as well?)
Also, from what I read [there](https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/glasgow_exts.html#unpack-pragma), it seams we need to add `-O` to the build for this to have any effect?

Reply:
I copied that from a haskell binary data tutorial, not sure if it makes any difference.